flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/IndexedInputGate.java

```diff
@@ -64,4 +64,8 @@ public void blockConsumption(InputChannelInfo channelInfo) {
     public void convertToPriorityEvent(int channelIndex, int sequenceNumber) throws IOException {
         getChannel(channelIndex).convertToPriorityEvent(sequenceNumber);
     }
+
+    public abstract int getBuffersInUseCount();
```

Comment:
It is actually a big question does it make sense to calculate the buffer size based on `buffers in use` or not. because as you can see the simple implementation requires extra synchronization + iteration over all channels which altogether is pretty expensive. It is possible to optimize this solution but it will lead to extra complexity. I still assume that it should help with data skew (at least it should be better than the total configured size)  but I don't sure that it is the best tradeoff between performance and implementation so we need to investigate it somehow.

Reply:
I can see that. Let me think about it tomorrow.

Reply:
On a second thought, I think this should be not worse compared to the `OutputFlusher`, which by default is executed every `100ms`. I would start with this, and if it turns out a problem, we can optimise it in the future.

Also we should check if it indeed solves the data skew case.